Add Batch data plane commands.

xingwu1 opened the pull request.

azurecla commented:
Hi __@xingwu1__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

troydai commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/commands.py:
```diff
@@ -5,7 +5,9 @@
-from ._client_factory import batch_client_factory
+from ._client_factory import batch_client_factory, batch_data_service_factory
+from ._command_type import cli_data_plane_command
+from ._validators import validate_pool_settings
```
It is recommended to use full path instead of relative path. We do have a lot of relative path import in existing code base. I think they're bad practice, too.

xingwu1 replied:
resolved.

troydai commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_command_type.py:
```diff
@@ -0,0 +1,748 @@
+# TODO: Enum choice lists
+
+
+_CLASS_NAME = re.compile(r"<(.*?)>")  # Strip model name from class docstring
```
Is there a way to avoid regular expression?

xingwu1 replied:
investigating.

xingwu1 commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/tests/data/batchUpdatePool.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "startTask": {
+    "commandLine": "cmd /c echo updated",
```
No, this is just a string to the test case, doesn't mean anything.

xingwu1 commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/custom.py:
```diff
@@ -90,11 +107,238 @@ def _upload_package_blob(package_file, url):
+    except (ValidationError, ClientRequestError) as ex:
+        raise CLIError(ex)
+
+resize_pool.__doc__ = PoolResizeParameter.__doc__
```
I couldn't figure out how to use double decroator wrappers (need pass the parameter to decrator) and keep the original function signature. Could you give me a hint or example?

troydai replied:
Sure. Here's an [example](https://github.com/Azure/azure-cli/blob/master/src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/custom.py#L40)

Python decorator is defined in [PEP 318](https://www.python.org/dev/peps/pep-0318/)

xingwu1 commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/custom.py:
```diff
@@ -90,11 +107,238 @@ def _upload_package_blob(package_file, url):
+        raise CLIError(ex)
+
+
+def resize_pool(client, pool_id, target_dedicated=None, #pylint:disable=too-many-arguments
```
Resolved.

troydai commented:
Note, if you want to check all PEP8 violation just run:
```
check_style --pep8 --module batch
```

troydai commented on src/azure-cli-core/azure/cli/core/decorators.py:
```diff
@@ -95,8 +95,10 @@ def _wrapped_func(*args, **kwargs):
 
 
-def transfer_doc(source_func):
+def transfer_doc(source_func, *additional_source_funcs):
```
The additional use case is a rare use case. Concatenating docstring is not use case in any other command module anytime soon. Therefore it is not ideal to modify the decorator just to fit a corner case. Please revoke the change and write a customized decorator in your own module.

xingwu1 replied:
resolve

troydai commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/custom.py:
```diff
@@ -265,8 +262,6 @@ def update_pool(client, pool_id, json_file=None, command_line=None,  # pylint:di
     except (ValidationError, ClientRequestError) as ex:
         raise CLIError(ex)
 
-update_pool.__doc__ = PoolUpdatePropertiesParameter.__doc__ + "\n" + StartTask.__doc__
```
Out of curiosity, why concatenate two docstrings? Isn't the format of the new doc string broken?

xingwu1 replied:
No one will intend to directly call the this custom function. The docstring here is only used for build up the cli param help string. So the broken doc string doesn't bother any real user.

annatisch commented:
@troydai,
The complex logic is almost entirely executed only on the specific command that is being run. Only one regex is executed in loading the module - and that's a [precompiled one](https://github.com/xingwu1/azure-cli/blob/master/src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_command_type.py#L37).
Unless the arguments for every command are loaded when the module is loaded? I was assuming the argument loader function was only called one that specific command had been executed?

troydai commented:
@annatisch thanks for the clarification. I'm glad to know that. However, I think you should probe what hit the performance during the loading. The `batch` command is two times slower than the second slowest command module in the list. It absolute time will only increase in the user environment.

annatisch commented:
@troydai,
That's for confirming that - yes I've been running it under the VS profiler, though having difficulties finding the culprit. Still investigating...

annatisch commented:
@troydai, additional - do you know what cmd I can use to specifically run and display those loading stats?

derekbekoe commented:
@annatisch `az --debug`

derekbekoe commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/custom.py:
```diff
@@ -90,11 +114,233 @@ def _upload_package_blob(package_file, url):
+        raise CLIError(ex)
+
+
+def delete_certificate(client, thumbprint, thumbprint_algorithm, abort=None):
```
Since in here, the action is `store_true`, `abort=False` makes more sense IMHO.

annatisch commented:
@troydai - thanks, I just figured that out ;)

derekbekoe commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/custom.py:
```diff
@@ -90,11 +114,233 @@ def _upload_package_blob(package_file, url):
+        except AttributeError:
+            raise CLIError(ex)
+    except (ValidationError, ClientRequestError) as ex:
+        raise CLIError(ex)
```
This try/except logic is used multiple times.
Consider moving out to helper function that takes a callable.
e.g.:
```
def delete_certificate(...):
    def handler():
        if abort:
             client.cancel_deletion(thumbprint_algorithm, thumbprint)
         else:
             client.delete(thumbprint_algorithm, thumbprint)
    _ex_handle_patch_exceptions(handler)

def _ex_handle_patch_exceptions(some_handler):
    try:
        some_handler()
    except BatchErrorException ...
       ...
```

Something like that.

xingwu1 replied:
resolved.

derekbekoe commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/custom.py:
```diff
@@ -90,11 +114,233 @@ def _upload_package_blob(package_file, url):
+                                          if_unmodified_since=if_unmodified_since)
+
+    try:
+        return client.resize(pool_id, param, pool_resize_options=resize_option)
```
I question this convention that I've seen in this PR of having an `abort` flag that executes a different operation.
Should consider having these as two separate commands.

Does `client.stop_resize` return the same object structure as `client.resize`?
If not, this is another reason why you may want these as two separate commands.

derekbekoe commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/custom.py:
```diff
@@ -90,11 +114,233 @@ def _upload_package_blob(package_file, url):
+
+
+@transfer_doc(PoolUpdatePropertiesParameter, StartTask)
+def update_pool(client, pool_id, json_file=None, command_line=None,  # pylint:disable=too-many-arguments, W0613
```
Replace `W0613` with the appropriate short-message.
This link should help http://pylint-messages.wikidot.com/all-messages.

derekbekoe replied:
Same with other instances.

derekbekoe commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/tests/data/batch-pool-create-invalid.json:
```diff
@@ -0,0 +1,16 @@
+    "osFamily": "4",
+    "targetOSVersion": "*"
+  }
+  "targetDedicated": 2,
```
This character is highlighted in red on GitHub which makes me suspicious.
Can you verify that it's a `:` (colon)?

troydai replied:
Missing a `,` following previous entity (on line 7).

annatisch replied:
@troydai, @derekbekoe,
This is a deliberately incorrect JSON file to test for bad input handling. Hence the file name suffix "invalid.json" :)

derekbekoe replied:
ahhh okay!

annatisch commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_client_factory.py:
```diff
@@ -5,7 +5,20 @@
 from azure.mgmt.batch import BatchManagementClient
 
+import azure.batch.batch_service_client as batch
+import azure.batch.batch_auth as batchauth
```
Oh good catch - thanks!
Seems like imports are one of our major loading perf issues at this point

annatisch commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_params.py:
```diff
@@ -8,28 +8,72 @@
+for command in ['list', 'show', 'create', 'set', 'delete', 'package']:
+    register_cli_argument('batch application {}'.format(command), 'account_name', batch_name_type, options_list=('--name', '-n'), validator=application_enabled)
+
+register_cli_argument('batch pool resize', 'if_modified_since', help='Specify this header to perform the operation only if the resource has been modified since the specified date/time.', type=datetime_format, arg_group='Pre-condition')
```
@derekbekoe - yeah these strings were copied straight from our swagger spec - we're endevouring to clean up this kind of terminology both here and in the spec itself :)

xingwu1 replied:
resolved.

annatisch commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_params.py:
```diff
@@ -8,28 +8,72 @@
+for item in ['batch certificate delete', 'batch certificate create', 'batch pool resize', 'batch pool reset', 'batch job list', 'batch task create']:
+    register_extra_cli_argument(item, 'account_name', arg_group='Batch Account',
+                                validator=validate_client_parameters,
+                                help='Batch account name. Environment variable: AZURE_BATCH_ACCOUNT')
```
@derekbekoe - These three environment variables are for authentication with our data plan API and are named as such to maintain consistency with the Xplat CLI.

xingwu1 replied:
resolve.

derekbekoe replied:
Consider using `az_config`, the CLI configuration feature for this.

This would allow the batch account to be set with `AZURE_BATCH_ACCOUNT` (as required) but also allow setting the config in the CLI configuration file.

For example https://github.com/Azure/azure-cli/blob/master/src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py#L23.

Another example https://github.com/Azure/azure-cli/blob/master/src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py#L51.

xingwu1 replied:
Already use az_config, if you look _validator.validate_client_parameters function.

tjprescott commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_client_factory.py:
```diff
@@ -5,7 +5,20 @@
+    account_name = kwargs.pop('account_name', None)
+    account_key = kwargs.pop('account_key', None)
+    account_endpoint = kwargs.pop('account_endpoint', None)
+    kwargs.pop('force', None)
```
What is this for?

tjprescott commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_command_type.py:
```diff
@@ -0,0 +1,750 @@
+        self.done = True
+
+
+class AzureDataPlaneCommand(object):
```
This should probably be AzureBatchDataPlaceCommand

tjprescott commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_command_type.py:
```diff
@@ -0,0 +1,750 @@
+                                             help=docstring))
+
+
+def cli_data_plane_command(name, operation, client_factory, transform=None,  # pylint:disable=too-many-arguments
```
This should be `cli_batch_data_plane_command`.

tjprescott commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_params.py:
```diff
@@ -8,28 +8,72 @@
+register_cli_argument('batch certificate', 'thumbprint', help='The certificate thumbprint.')
+register_cli_argument('batch certificate', 'thumbprint_algorithm', help='The certificate thumbprint algorithm.')
+register_cli_argument('batch certificate', 'password', help='The password to access the certificate\'s private key.')
+register_cli_argument('batch certificate', 'cert_file', type=file_type, help='The certificate file: cer file or pfx file.', validator=validate_cert_file, completer=FilesCompleter())
```
You should also specify `type=file_type` here and import file_type from `azure.cli.core.commands.parameters`. This ensures the leading ~ is expanded correctly.

xingwu1 replied:
It has already set the type to file_type.

tjprescott commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_params.py:
```diff
@@ -8,28 +8,72 @@
+register_cli_argument('batch certificate', 'cert_file', type=file_type, help='The certificate file: cer file or pfx file.', validator=validate_cert_file, completer=FilesCompleter())
+register_cli_argument('batch certificate delete', 'abort', action='store_true', help='Cancel the failed certificate deletion operation.')
+
+register_cli_argument('batch task create', 'json_file', type=file_type, help='The file containing the task(s) to create in JSON format, if this parameter is specified, all other parameters are ignored.', validator=validate_json_file, completer=FilesCompleter())
```
Same here. `type=file_type`. Also, an alias to `--task-file` might be more intuitive than "--json-file".

xingwu1 replied:
It already be file_type type. I will udpate the name.

tjprescott commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_params.py:
```diff
@@ -8,28 +8,72 @@
+                                help='Batch account name. Environment variable: AZURE_BATCH_ACCOUNT')
+    register_extra_cli_argument(item, 'account_key', arg_group='Batch Account',
+                                help='Batch account key. Must be used in conjunction with Batch account name and endpoint. Environment variable: AZURE_BATCH_ACCESS_KEY')
+    register_extra_cli_argument(item, 'account_endpoint', arg_group='Batch Account',
```
I assumed you added these three arguments to the Azure{Batch}DataPlaneCommand object to avoid having to do this. Why do you do both?

xingwu1 replied:
The AzureDataPlaneCommand is automatic create the parameter, and here is manually added to the only 5 custom commands.

xingwu1 commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_validators.py:
```diff
@@ -3,20 +3,236 @@
+try:
+    from urllib.parse import urlsplit
+except ImportError:
+    from urlparse import urlsplit  # pylint: disable=import-error
```
resolved.

xingwu1 commented:
@troydai  @derekbekoe 

The latest result show the loading performance has been improved a lot

DEBUG: Loaded module 'batch' in 0.056 seconds

Consider the batch service has around 90 commands, it is quite good result.

tjprescott commented:
Also, one thing we require of all New Command PRs is a dump of the -h help text. This helps us spot things like missing or illogical parameter help. Not every command is necessary--generally just the create commands. 

Please add those as a comment on the PR (and omit the global arguments please).

annatisch commented:
@tjprescott,
I've generated the help text into a separate branch on my fork:
[Summary Commands](https://github.com/annatisch/azure-cli/blob/help_docs/batch_short_help.txt) (includes all create commands, both custom and 'auto parsed' and a couple of others, no global params)
[Full Commands](https://github.com/annatisch/azure-cli/blob/help_docs/batch_help_full.txt) (includes all commands and globals)

tjprescott commented:
@annatisch is the help text updated? I started looking at it and immediately saw:
```
    --storage-account-id          : The resource id of the storage account to be used for auto

                                    storage account.

    --storage-account-name        : The storage account name to be used for auto storage account.
```

This is not consistent with the rest of the CLI. It would be called `storage_account` which would accept a name or id. But in the custom command it only looks like `storage_account_id`. 

I just want to make sure before I start commenting on the help text that it's the latest.

annatisch commented:
@tjprescott,
It was generated yesterday - so while it's not 100% latest, it's pretty recent.
Also - the command arguments you mentioned there are actually from the management module which has already been merged into master. But we can update it :)

FYI @xingwu1

xingwu1 commented:
@tjprescott,
I change the code to use single storage account to accept both name and id scenario.

annatisch commented:
@tjprescott,
I've updated the help text based on Xing's update above and an internal review.

tjprescott commented:
For `az batch application create`: 
- recommend for `--allow-updates`: "Specify to indicate that packages within the application may be overwritten using the same version string."  That's more in line with how the rest of the CLI documents flags.
- For display-name, it might be helpful to mention what, if anything, is used if this is not specified, since it is optional. Will it just show the application-id?

tjprescott commented:
For `az batch application package create`:
- Is `--version` a number of version string? It might be helpful for point out. For example, IIRC Android applications have two versions: a version number that must always increase and then a version string which is for display purposes only.

tjprescott commented:
For `az batch certificate create`:
Is there an enum for what options are allowed for `--thumbprint-algorithm`? If so, it would be good to use the `**enum_choice_list` method for that parameter so that users know which algorithms are supported. This can be used even if there isn't an enum model in the SDK to reflect on. You can use `**enum_choice_list(['foo', 'doo', 'bar'])` to specify a custom list and get the benefits enum_choice_list confers.

tjprescott commented:
My general feedback on some of the other creates (the three or four super long ones):

- It seems like there are things which are or should be enums and thus exposed as choice lists (for example, os-family).
- in general, where there is a parameter called `--{foo}-id`, if that ID is an ARM resource ID, then that parameter should just be called `{foo}` and it should accept a name or ID.
- you might be able to shorten some of the options since you have them in argument groups. For example: the Job Manager Task arguments all start with --job-manager-task. If you can remove that and not have any naming conflicts, it might be worth considering.

tjprescott commented:
Do you have data from the other CLIs on what the distribution is between customers that use the "file" option to specify everything in a JSON file vice those who actually use the 30+ arguments?

To me, the sheer number of arguments possible for some of the creates is overwhelming. If you had data on which ones (if any) could be suppressed or deferred to an update step, you could remove or default them to make the usability easier. You can always add those arguments later if people are demanding them, but removing arguments later will be much harder.

@johanste do you have thoughts/advice on this?

xingwu1 commented:
@tjprescott 

Here is summary of answer to your questions:
1. --allow-updates: we don't want to change the description to "specify to ...", this string is directly copied from python SDK and come from swagger spec. We removed the "specify to ..." string from a property of swagger spec since C#/Java SDK will automatic add "Gets or sets... " to the beginning of the property title string. Right now, the properties title in our swagger spec all starts with just noun instead of verb.

2. "dispaly-name" is just a FYI string, if you don't specify it, it just return none.
3. --version is like a id, it can be anything string (not necessary as version). This is bad naming of our swagger spec (the root cause is our RP API). We don't want to do anything from our cli unless the RP side change the string.
4. we will add the enum_choice_list to "thumbprint-algorithm"
5. we will add the enum_choice_list(Type) to possible enum parameters, also fix the os-family list
6. The only string fit into ARM id category is "storage-account-id", which I already changed to "storage-account" parameter.
7. we only add prefix to parameter when we find there are conflict between 2 parameter. For example, both "job-manager-task" and "job-prepartion-task" have property named "id", to flatten them in the command line parameter, we rename them to "job-manager-task-id" and "job-prepartion-task-id", otherwise, the parameter names are exactly same as properties name.
8. we will do more investigate to reduce the level of flatten to reduce the parameter number.

annatisch commented:
@tjprescott, @troydai 

We've just finished an internal review of the commands and help documentation and I have updated the outputs here:
[Summary Commands ](https://github.com/annatisch/azure-cli/blob/help_docs/batch_short_help.txt)(includes all create commands, both custom and 'auto parsed' and a couple of others, no global params)
[Full Commands ](https://github.com/annatisch/azure-cli/blob/help_docs/batch_help_full.txt)(includes all commands, group headings and globals parameters)

The commands have reduced quite significantly - though still quite numerous. Let us know what you think. We've addressed the enum options (and removed thumbprint-algorithm), and prefixed more commands for clarification. We've reworded quite a lot of the help text to be more informative (e.g. correct use of Boolean flags).

FYI @itowlson

xingwu1 commented:
@tjprescott 
We reduced more parameters and add the VM image completer.

xingwu1 commented:
@troydai Could you review the change and sign off it if you feel good with the change?

annatisch commented:
@troydai - would be great if you could take a look at the changes so we can get this merged today :)

troydai commented:
I'm taking another look.

troydai commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_validators.py:
```diff
@@ -3,20 +3,309 @@
+    except DeserializationError:
+        message = "Argument {} is not a valid ISO-8601 datetime format"
+        raise ValueError(message.format(value))
+    else:
```
not sure if this `else` is necessary here.

troydai commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/_validators.py:
```diff
@@ -3,20 +3,309 @@
+def validate_pool_resize_parameters(namespace):
+    """Validate pool resize parameters correct"""
+    if not namespace.abort:
+        if not namespace.target_dedicated:
```
`if not namespace.abort and not namespace.target_dedicated:`

johanste commented:
Thanks everybody! Great to see this merged!

annatisch commented:
Thanks guys!!

xingwu1 commented:
Thanks!!!
